Fast and Flexible GPU Quantization for Transformers.md

```diff
@@ -1,82 +1,193 @@
+
+This has led to a surge in interest for optimizing transformer inference. There are a range of techniques available [1], including:
+
+* Sparsity
+* Pruning
+* Conditional Computation (e.g., Mixture of Experts)
+* Knowledge Distillation
+* Quantization
+
+Of these, quantization is in some sense the most universal. It can be applied to any network, regardless of architecture.
+
+By reducing the precision of model parameters and activations, quantization aims to minimize increase throughput and decrease memory footprint, at the cost of potentially damaging model accuracy.
+
+Provided the decrease in accuracy is minimal, this sounds ideal. However, this research area is still young, and implementing a GPU-based quantization scheme that *actually speeds up your model* is not straightforward. The main challenges stem from a lack of fast General Matrix Multiply (GEMM) GPU kernels for precisions lower than INT8, and the overheads associated with converting between floating point (FP) and Integer (INT) data types.
+
+In this blog, we provide a detailed guide to GPU-based quantization of transformers. We describe an approach that is both flexible and capable of genuinely improve throughput. The content is organized as follows:
+
+* [Background](#Background)
+	* [The Quantization Equation](#The%20Quantization%20Equation)
+	* [Dynamic vs Static Quantization](#Dynamic%20vs%20Static%20Quantization)
+	* [Quantization Granularity](#Quantization%20Granularity)
+* [Important Concepts](#Important%20Concepts)
+	* [Calibration](#Calibration)
+	* [Specifics of INT8 GEMMs](#Specifics%20of%20INT8%20GEMMs)
+		* [i8f16](#i8f16)
+		* [i8i8](#i8i8)
+		* [Quantization Operation Overheads](#Quantization%20Operation%20Overheads)
+	* [SmoothQuant](#SmoothQuant)
+* [Implementation](#Implementation)
+	* [GPU Quantization in Practice](#GPU%20Quantization%20in%20Practice)
+	* [Memory Layouts](#Memory%20Layouts)
+	* [Fusion Strategy (and diagrams)](#Fusion%20Strategy%20(and%20diagrams))
+* [Some Brief Results](#Some%20Brief%20Results)
+	* [Accuracy](#Accuracy)
+	* [Throughput](#%20Throughput)
+* [References](#References)
+
 
-Disclaimer that this is _scalar_ quantization, not _vector_ quantization.
 
 ## Background
 
-We’ll begin with a high-level summary of quantization. If you’d like more reading on this topic, we’ve listed some nice blogs/papers in the References section [1-4].
+We’ll begin with a quick summary of quantization. For further reading on this subject, we’ve listed some nice blogs/papers in our [References](#References) section [1-4].
 
 ### The Quantization Equation
-In principle, we can use any function to map between floating point (FP) and integer (INT) values. But it’s simplest (and quickest on hardware) to use a linear operation [4]:
+In principle, we can use any function to map between FP and INT values. But a linear function is simplest and quickest on the hardware [4]:
```

Comment:
I guess the linear function is about scaling the range, so it's less about FP -> INT and more about quantization in general (e.g. FP32 -> FP16 is still quantization)

Reply:
Have made this a bit more general